core/src/main/java/org/gbif/pipelines/core/parsers/location/LocationMatcher.java

```diff
@@ -0,0 +1,177 @@
+      }
+      return ParsedField.success(ParsedLocation.newBuilder().country(countryMatch.get()).latLng(latLng).build(),
+                                 issues);
+    }
+
+    if (this.country != null) {
+      // try alternative transformations if the country was supplied
+      LOG.info("Trying alternative transformations to the coordinates");
+      for (CoordinatesFunction transformation : alternativeTransformations) {
+        // transform location
+        LatLng latLngTransformed = transformation.getTransformer().apply(this.latLng);
+
+        // call ws
+        Optional<List<Country>> countriesOpt = getCountriesFromCoordinates(latLngTransformed);
+        if (!countriesFound(countriesOpt)) {
+          continue;
+        }
+
+        // match with the country supplied
+        countryMatch = matchWithCountry(countriesOpt.get());
+        if (countryMatch.isPresent()) {
+          // country found
+          // add issue from the transformation
+          CoordinatesFunction.getIssueTypes(transformation)
+            .forEach(issueType -> issues.add(new InterpretationIssue(issueType, getCountryAndCoordinatesTerms())));
+          return ParsedField.success(ParsedLocation.newBuilder()
+                                       .country(countryMatch.get())
+                                       .latLng(latLngTransformed)
+                                       .build(), issues);
+        }
+      }
+    }
+
+    // no result found
+    issues.add(new InterpretationIssue(IssueType.COUNTRY_COORDINATE_MISMATCH, getCountryAndCoordinatesTerms()));
+    return ParsedField.fail(issues);
+  }
+
+  private Optional<Country> tryIdentityMatch(List<InterpretationIssue> issues) {
+    // call WS with identity coords
+    Optional<List<Country>> countriesOpt = getCountriesFromCoordinates(latLng);
+
+    if (!countriesFound(countriesOpt)) {
+      return Optional.empty();
+    }
+
+    List<Country> countries = countriesOpt.get();
+    // if not country supplied
+    if (country == null) {
+      return Optional.ofNullable(countries.get(0));
+    }
+
+    // else try toi match with the country supplied
+    Optional<Country> countryMatched = matchWithCountry(countries);
+
+    if (countryMatched.isPresent()) {
+      return countryMatched;
+    }
+
+    // if we have not found it yet, we try with equivalent countries
+    Optional<Country> countryMatch = containsAnyCountry(CountryMaps.equivalent(country), countries);
+    if (countryMatch.isPresent()) {
+      return countryMatch;
+    }
+
+    // if we have not found it yet, we try with confused countries
+    countryMatch = containsAnyCountry(CountryMaps.confused(country), countries);
+    if (countryMatch.isPresent()) {
+      issues.add(new InterpretationIssue(IssueType.COUNTRY_DERIVED_FROM_COORDINATES, DwcTerm.country));
+      return countryMatch;
+    }
+
+    return Optional.empty();
+  }
+
+  private Optional<List<Country>> getCountriesFromCoordinates(LatLng latLng) {
```

Comment:
this method can be changed to return Optional.empty() when the list of countries in the response is empty, or we can remove the the usage of Optionals and user a List, int that way we can make the code more compact and probably remove the method 'countriesFound'